### PR TITLE
New lexer: lustre

### PR DESCRIPTION
**Pull request by jahierwan**

### CODE REVIEW

**Comment by jahierwan** [comment]
I've also put the lutin lexer in this commit as lustre and lutin are both new and similar

**Comment by jahierwan** [comment]
up ?

File: lib/rouge/demos/lutin [hunk 1]
```diff
@@ -0,0 +1,84 @@
+-- The examples provided in [SIES'13] "Engineering Functional
```
Comment (vidarh) [inline lib/rouge/demos/lutin]: It'd be good to trim down the demos/ files a bit. Most are in the 10-30 line length, as they're meant to be able to give a tiny showcase such as e.g. in the boxes here: http://rouge.jneen.net/

Comment (jahierwan) [reply lib/rouge/demos/lutin]: Sure, no pb. There are 2 sets of demos :
- lib/rouge/demos/
- spec/visual/samples/

Do you want me to reduce both?

Comment (jahierwan) [reply lib/rouge/demos/lutin]: Ok. So I will probably leave it as it is, so that the example compiles.

Thanks for taking care of this.

**Comment by jahierwan** [comment]
is there something i can do what this ci failure ?

**Comment by pyrmont** [comment]
@jahierwan I'm sorry it's taken such a long time to get to this submission >_<

I've rebased your branch against the most up to date version of master and force pushed the code to your fork. I can see the logic for why you've combined the Lutin and Lustre lexers but could I trouble you to create a branch off this and pull them apart? They don't share any code and should be submitted separately.

I tried to do it for you but unfortunately I only have commit rights to the branch you've used to submit the PR; I can't create a new branch on your fork.

File: lib/rouge/demos/lustre [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	s = if onelatetrain then false else ss;
+tel
+	
+node metro (sec: bool; beacons: bool^nbtrains)
+returns (ontime, onbrake, late,  stopped: bool^nbtrains);
+var second : bool;
+let
+	second = clock(late, sec);
+	(ontime, onbrake, late,  stopped) = map<<train;nbtrains>> (sec^nbtrains, beacons);
+tel
```
Comment (pyrmont) [inline lib/rouge/demos/lustre]: Could you cut down this example? You only need a very short demo (4-5 lines is plenty).

File: lib/rouge/lexers/lustre.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      def self.primitives
+        @primitives ||= Set.new %w(int real bool)
+      end
+
+      operator = %r([;,_!$%&*+./:<=>?@^|~#-]+)
+      id = /[a-z_][\w']*/i
+
+      state :root do
+        rule /\s+/m, Text
+        rule /false|true|[(][)]|\[\]/, Name::Builtin::Pseudo
```
Comment (pyrmont) [inline lib/rouge/lexers/lustre.rb]: This doesn't look like the right tokens to use for these values. `true` and `false` should be `Keyword::Constant`. As for `()` and `[]`, these are normally just left to be highlighted as punctuation (you have a rule later that will do that).

File: lib/rouge/lexers/lustre.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        rule %r(\-\-.*?$), Comment::Single
```
Comment (pyrmont) [inline lib/rouge/lexers/lustre.rb]: You can just express the pattern `.*`. There's no need to make it lazy or worry about matching to the end of the line (Ruby's wildcards don't match newlines unless the `m` option is set).

File: lib/rouge/lexers/lustre.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+          elsif self.class.word_operators.include? match
+            token Operator::Word
+          elsif self.class.primitives.include? match
+            token Keyword::Type
+          else
+            token Name
+          end
+        end
+
+        rule /[(){}\[\];]+/, Punctuation
```
Comment (pyrmont) [inline lib/rouge/lexers/lustre.rb]: You have as both punctuation and as an operator. The operator token will never match. Is that OK?

File: lib/rouge/lexers/lustre.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        rule /-?\d[\d_]*(.[\d_]*)?(e[+-]?\d[\d_]*)/i, Num::Float
+        rule /0x\h[\h_]*/i, Num::Hex
+        rule /0o[0-7][0-7_]*/i, Num::Oct
+        rule /0b[01][01_]*/i, Num::Bin
+        rule /\d[\d_]*/, Num::Integer
+
+        rule /'(?:(\\[\\"'ntbr ])|(\\[0-9]{3})|(\\x\h{2}))'/, Str::Char
+        rule /'[.]'/, Str::Char
+        rule /'/, Keyword
```
Comment (pyrmont) [inline lib/rouge/lexers/lustre.rb]: The apostrophe by itself is a keyword?

File: lib/rouge/lexers/lutin.rb [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        rule /"/, Str::Double, :pop!
+      end
+
+      state :escape_sequence do
+        rule /\\[\\"'ntbr]/, Str::Escape
+        rule /\\\d{3}/, Str::Escape
+        rule /\\x\h{2}/, Str::Escape
+      end
+
+      state :dotted do
```
Comment (pyrmont) [inline lib/rouge/lexers/lutin.rb]: This state isn't pushed onto the stack or mixed in. Is that an oversight or is it from an older version of the code and can now be deleted?

File: spec/visual/samples/lustre [hunk 8]
```diff
@@ -0,0 +1,244 @@
+
+-------------------------------------------------------------------
+(* Compute the roof motion speed *)
+  
+const period = 0.1;  -- in seconds
+const wheel_girth = 1.4;  -- meter
+const size = 50; -- size of the sliding windows used to smooth values
+const max_roof_speed = 10.0;
+const speed_max = 110.0;                                           --
+const speed_hist= 5.0;
```
Comment (pyrmont) [inline spec/visual/samples/lustre]: Could you replace some of these floats with binary, octal and hexadecimal numbers to ensure they're lexing correctly?

File: spec/visual/samples/lustre [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+         (locked    -> if (OnOff and Parked) when locked(pst) then in_motion else locked)
+         (in_motion -> if Done when in_motion(pst) then locked else in_motion);
+  Locked = (st=locked) ;
+  Tick_on_in_motion = Tick when in_motion(st);
+  Roof_Speed = merge st
+    (locked    -> 0.0 when locked(st))
+    (in_motion -> roof_speed(Tick_on_in_motion));
+tel
+
+type roof_speed_state = enum { wait, fast, slow };
```
Comment (pyrmont) [inline spec/visual/samples/lustre]: Could you delete the text that's duplicative of earlier code in the sample? This will help in the future when working on the lexer and ensuring there is adequate coverage in the sample.

File: lib/rouge/demos/lustre (hunk 2 above)
Comment (jahierwan) [reply lib/rouge/demos/lustre]: Sure.

File: lib/rouge/lexers/lustre.rb (hunk 3 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: Ok.
Basically, I've taken inspiration from ocaml.rb, since ocaml is a langage I know.

File: lib/rouge/lexers/lustre.rb (hunk 4 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: Ok.

File: lib/rouge/lexers/lustre.rb (hunk 5 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: I will remove the line

File: lib/rouge/lexers/lustre.rb (hunk 6 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: No. I will remove it.

File: spec/visual/samples/lustre (hunk 9 above)
Comment (jahierwan) [reply spec/visual/samples/lustre]: I'm not sure why you say it's a duplicate. If I remove the definition of roof_speed_state type,
It won't be a working example for sure (afa the lustre compiler is concerned.

File: spec/visual/samples/lustre (hunk 8 above)
Comment (jahierwan) [reply spec/visual/samples/lustre]: Octal and hexadecimal numbers are (currently) not supported by the Lustre compiler (maybe they should).

Comment (jahierwan) [reply spec/visual/samples/lustre]: Ok, I will remove them from the syntax

File: lib/rouge/lexers/lustre.rb (hunk 5 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: Actually removing that line broke test. I've put it back.
Could you pleasereword the sentence "You have as both punctuation and as an operator" ?
I don't understand it.

File: lib/rouge/lexers/lutin.rb (hunk 7 above)
Comment (jahierwan) [reply lib/rouge/lexers/lutin.rb]: I'll adress this in the lutin dedicated PR.

Comment (jahierwan) [reply lib/rouge/lexers/lutin.rb]: cf PR#58

File: lib/rouge/lexers/lustre.rb (hunk 5 above)
Comment (pyrmont) [reply lib/rouge/lexers/lustre.rb]: Sorry—the initial comment was poorly worded >_<

What I was trying to say was that this line and the line that comes after it (`rule operator, Operator`) both match against `;`. The comment meant to ask: how should `;` be lexed? Tokenising it as `Punctuation` looks correct to me but if that's the case, it should be removed from the definition of `operator`.

Speaking of which, should some of the other characters be moved out of the `operator` rule? Is `,` or `_` really an operator?

File: lib/rouge/lexers/lustre.rb [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+      title "Lustre"
+      desc 'The Lustre programming language (Verimag)'
+      tag 'lustre'
+      filenames '*.lus'
+      mimetypes 'text/x-lustre'
+
+      def self.keywords
+        @keywords ||= Set.new %w(
+          extern unsafe assert const current enum function
+          false let node operator returns
```
Comment (pyrmont) [inline lib/rouge/lexers/lustre.rb]: Sorry, I missed this in the initial review but `false` and `true` should be removed from this list.

File: lib/rouge/lexers/lutin.rb (hunk 7 above)
Comment (pyrmont) [reply lib/rouge/lexers/lutin.rb]: I'm a bit confused—this state is still in the lexer and is still never used. it should be deleted, right?

File: spec/visual/samples/lustre (hunk 9 above)
Comment (pyrmont) [reply spec/visual/samples/lustre]: I'm sorry—again, poorly worded on my part.

The comment was meant as a comment on the visual sample generally, not this specific line. A large portion of the visual sample looks to me to check text that doesn't test anything in the lexer that an earlier line didn't test. Some duplication is fine but it's preferable to have a visual sample that more clearly tests the various rules in the lexer. That way, it's easy to see if there's a test of each rule and where to add lines to the sample over time as bugs and edge cases are discovered.

File: lib/rouge/lexers/lustre.rb (hunk 5 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: ok, thnaks for the clarification. I'll remove ';' from the definition of operator.

Concerning ',' and '_',  maybe not. Once again, I've taken inspiration from ocaml.rb, from which the role of ',' '_', or ';' is similar than in Lustre to some extent. In Lustre clearly, ';' is a punctuation (in ocaml too). ',' appears in structures and arrays, and it may make sense to see them as operator. '_' is an identifier as in ocaml; i'll remove it from the operator section.

File: lib/rouge/lexers/lutin.rb (hunk 7 above)
Comment (jahierwan) [reply lib/rouge/lexers/lutin.rb]: Why do you say that? I've removed it.
https://github.com/rouge-ruby/rouge/pull/1307/commits/7c62be91edf84b8cf4f5a710495d35dbdcdfdd56

File: spec/visual/samples/lustre (hunk 9 above)
Comment (jahierwan) [reply spec/visual/samples/lustre]: Ok. So it's ok if my example does not compile because some definitions are missing?

File: lib/rouge/lexers/lustre.rb (hunk 10 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: ok.

File: lib/rouge/lexers/lutin.rb (hunk 7 above)
Comment (pyrmont) [reply lib/rouge/lexers/lutin.rb]: I'm talking about the `:dotted` state in [the Lustre lexer](https://github.com/jahierwan/rouge/blob/3ee37a619c53810217e043a965820610ba505383/lib/rouge/lexers/lustre.rb#L78). It's still there and isn't ever used.

Comment (jahierwan) [reply lib/rouge/lexers/lutin.rb]: oh, ok. I did not understand in the first place because this is a comment of lutin.rb. I will remove it from lustre.rb too, thanks.

File: lib/rouge/lexers/lustre.rb (hunk 5 above)
Comment (pyrmont) [reply lib/rouge/lexers/lustre.rb]: There's some value in being consistent with previous lexers and if Lustre works similarly to OCaml then maintaining consistency makes sense. I'm not very familiar with OCaml but it does seem that `;` and `,` are referred to as operators in documentation ([example](https://ocaml.org/learn/tutorials/structure_of_ocaml_programs.html)) so if it's performing a similar function in Lustre, putting it back seems fine to me.

File: spec/visual/samples/lustre (hunk 9 above)
Comment (pyrmont) [reply spec/visual/samples/lustre]: Yeah, the sample doesn't need to be compilable code. Its purpose is to provide a visual way to check lexing. In your branch, if you run `rackup`, you should start a local server that you can visit to see what the lexed version of your sample will look like.

File: lib/rouge/lexers/lustre.rb (hunk 5 above)
Comment (jahierwan) [reply lib/rouge/lexers/lustre.rb]: Well, I'am not sure seing ";" as an operator in Lustre makes sense as it does in ocaml. Therefore I propose to leave it as it is currently.

File: spec/visual/samples/lustre (hunk 9 above)
Comment (jahierwan) [reply spec/visual/samples/lustre]: ok, so I've shortened it.

Comment (pyrmont) [reply spec/visual/samples/lustre]: It still seemed to have a lot of duplication. I've reorganised it a little and pushed an updated version to your branch. Let me know if it looks good to go on your end.

Comment (jahierwan) [reply spec/visual/samples/lustre]: It's fine. Thanks a lot.
